File: src/main/cpp/subsystems/drive/trajectory/GenerateTrajectories.cpp
// Copyright (c) 2025 FRC 6328
// http://github.com/Mechanical-Advantage
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file at
// the root directory of this project.

#include "org/littletonrobotics/frc2025/subsystems/drive/trajectory/GenerateTrajectories.h"
#include "absl/hash/hash.h"
#include "absl/hash/internal/city.h"
#include "absl/strings/ascii.h"
#include "absl/strings/escaping.h"
#include "absl/strings/match.h"
#include "absl/strings/numbers.h"
#include "absl/strings/str_cat.h"
#include "absl/strings/str_format.h"
#include "absl/strings/str_join.h"
#include "absl/strings/str_split.h"
#include "absl/strings/string_view.h"
#include "frc/Units/Units.h"
#include "google/protobuf/io/zero_copy_stream_impl.h"
#include "google/protobuf/text_format.h"
#include "google/protobuf/util/json_util.h"
#include "google/protobuf/util/message_differencer.h"
#include "org/littletonrobotics/frc2025/Constants.h"
#include "org/littletonrobotics/frc2025/subsystems/drive/DriveConstants.h"
#include "org/littletonrobotics/frc2025/subsystems/drive/trajectory/DriveTrajectories.h"
#include "org/littletonrobotics/vehicletrajectoryservice/VehicleTrajectoryService.grpc.h"
#include "org/littletonrobotics/vehicletrajectoryservice/VehicleTrajectoryService.grpc.pb.h"
#include <cmath>
#include <filesystem>
#include <fstream>
#include <google/protobuf/io/zero_copy_stream_impl.h>
#include <google/protobuf/text_format.h>
#include <google/protobuf/util/json_util.h>
#include <google/protobuf/util/message_differencer.h>
#include <grpcpp/create_channel.h>
#include <grpcpp/grpcpp.h>
#include <grpcpp/security/credentials.h>
#include <iomanip>
#include <iostream>
#include <sstream>

std::string
GenerateTrajectories::GetHashCode(const VehicleModel &model,
                                  const std::vector<PathSegment> &segments) {
  std::stringstream hashString;
  hashString.precision(6);
  hashString << std::fixed;

  hashString << model.mass();
  hashString << model.moi();
  hashString << model.vehicle_length();
  hashString << model.vehicle_width();
  hashString << model.wheel_radius();
  hashString << model.max_wheel_torque();
  hashString << model.max_wheel_omega();

  for (const auto &segment : segments) {
    for (const auto &waypoint : segment.waypoints()) {
      hashString << waypoint.x();
      hashString << waypoint.y();
      if (waypoint.has_heading_constraint()) {
        hashString << waypoint.heading_constraint();
      }

      if (waypoint.has_samples()) {
        hashString << waypoint.samples();
      }

      switch (waypoint.velocity_constraint_case()) {
      case Waypoint::VelocityConstraintCase::kZeroVelocity: {
        hashString << 0.0;
        break;
      }
      case Waypoint::VelocityConstraintCase::kVehicleVelocity: {
        hashString << waypoint.vehicle_velocity().vx();
        hashString << waypoint.vehicle_velocity().vy();
        hashString << waypoint.vehicle_velocity().omega();
        break;
      }
      case Waypoint::VelocityConstraintCase::VELOCITY_CONSTRAINT_NOT_SET: {
        break;
      }
      }
    }

    if (segment.has_max_velocity()) {
      hashString << segment.max_velocity();
    }
    if (segment.has_max_omega()) {
      hashString << segment.max_omega();
    }

    hashString << segment.straight_line();
  }

  return absl::StrCat(absl::Hash(hashString.str()));
}

int main(int argc, char **argv) {
  org::littletonrobotics::frc2025::Constants::DisableHAL();

  // Create vehicle model
  VehicleModel model;
  model.set_mass(67.0);
  model.set_moi(5.8);
  model.set_vehicle_length(org::littletonrobotics::frc2025::subsystems::drive::
                               DriveConstants::trackWidthX);
  model.set_vehicle_width(org::littletonrobotics::frc2025::subsystems::drive::
                              DriveConstants::trackWidthY);
  model.set_wheel_radius(frc::Units::InchesToMeters(2.0));
  model.set_max_wheel_torque(3.0);
  model.set_max_wheel_omega(org::littletonrobotics::frc2025::subsystems::drive::
                                DriveConstants::maxLinearSpeed /
                            frc::Units::InchesToMeters(2.0) * 0.75);

  // Connect to service
  auto channel = grpc::CreateChannel("127.0.0.1:56328",
                                     grpc::InsecureChannelCredentials());
  std::unique_ptr<VehicleTrajectoryService::Stub> service(
      VehicleTrajectoryService::NewStub(channel));

  std::set<std::string> completedPaths;
  while (true) {
    // Get all paths
    auto allPaths = org::littletonrobotics::frc2025::subsystems::drive::
        trajectory::DriveTrajectories::paths;
    bool hasAllPaths = true;
    for (const auto &supplier : org::littletonrobotics::frc2025::subsystems::
             drive::trajectory::DriveTrajectories::suppliedPaths) {
      auto suppliedPaths = supplier(completedPaths);
      if (suppliedPaths.empty() && !supplier) {
        hasAllPaths = false;
      } else {
        allPaths.insert(suppliedPaths.begin(), suppliedPaths.end());
      }
    }
    std::set<std::string> originalKeys;
    for (const auto &pair : allPaths) {
      originalKeys.insert(pair.first);
    }
    for (const auto &name : originalKeys) {
      if (completedPaths.count(name)) {
        allPaths.erase(name);
      } else {
        completedPaths.insert(name);
      }
    }
    if (!hasAllPaths && allPaths.empty()) {
      std::cerr << "Invalid dependency tree. Not all supplied trajectories are "
                   "available, but there are no trajectories to generate."
                << std::endl;
      return 1;
    }

    // Check hashcodes
    std::map<std::string, std::vector<PathSegment>> pathQueue;
    for (const auto &entry : allPaths) {
      std::string hashCode =
          GenerateTrajectories::GetHashCode(model, entry.second);
      std::filesystem::path pathFile("src/main/deploy/trajectories/" +
                                     entry.first + ".pathblob");
      if (std::filesystem::exists(pathFile)) {
        try {
          std::ifstream fileStream(pathFile, std::ios::binary);
          Trajectory trajectory;
          if (trajectory.ParseFromIstream(&fileStream)) {
            if (trajectory.hash_code() != hashCode) {
              pathQueue[entry.first] = entry.second;
            }
          } else {
            pathQueue[entry.first] = entry.second;
          }
        } catch (const std::exception &e) {
          std::cerr << "Error reading file: " << e.what() << std::endl;
          pathQueue[entry.first] = entry.second;
        }
      } else {
        pathQueue[entry.first] = entry.second;
      }
    }

    // Generate trajectories
    const char *generateEmptyFlag =
        std::getenv("GENERATE_EMPTY_DRIVE_TRAJECTORIES");
    bool generateEmpty =
        generateEmptyFlag != nullptr && std::strlen(generateEmptyFlag) > 0;
    for (const auto &entry : pathQueue) {
      Trajectory trajectory;
      std::cout << entry.first << " - Generating 💭";
      auto startTime = std::chrono::high_resolution_clock::now();
      if (generateEmpty) {
        trajectory.mutable_states()->Add();
      } else {
        // Use service for generation
        PathRequest request;
        *request.mutable_model() = model;
        for (const auto &segment : entry.second) {
          *request.add_segments() = segment;
        }

        TrajectoryResponse response;
        grpc::ClientContext context;
        grpc::Status status =
            service->GenerateTrajectory(&context, request, &response);
        if (!status.ok()) {
          std::cerr << "gRPC call failed: " << status.error_message()
                    << std::endl;
          return 1;
        }
        if (response.has_error() && !response.error().reason().empty()) {
          std::cerr << "Got error response for trajectory \"" << entry.first
                    << "\": " << response.error().reason() << std::endl;
          return 1;
        }
        trajectory = response.trajectory();
        trajectory.set_hash_code(
            GenerateTrajectories::GetHashCode(model, entry.second));
      }
      std::filesystem::path pathFile("src/main/deploy/trajectories/" +
                                     entry.first + ".pathblob");
      try {
        std::ofstream fileStream(pathFile, std::ios::binary);
        trajectory.SerializeToOstream(&fileStream);
        auto endTime = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(
                            endTime - startTime)
                            .count();
        std::cout << "\r" << entry.first << " - Finished in "
                  << std::round(duration / 100.0) / 10.0 << " secs ✅"
                  << std::endl;
      } catch (const std::exception &e) {
        std::cout << "\r" << entry.first << " - FAILED ⛔" << std::endl;
        std::cerr << "Error writing file: " << e.what() << std::endl;
      }
    }

    // Exit if all trajectories ready
    if (hasAllPaths) {
      break;
    }
  }

  // Delete old trajectories
  try {
    for (const auto &entry :
         std::filesystem::directory_iterator("src/main/deploy/trajectories")) {
      std::string filename = entry.path().filename().string();
      if (filename.find(".pathblob") == std::string::npos)
        continue;
      std::vector<std::string> components;
      absl::StrSplit(filename, '.', std::back_inserter(components));
      if (components.size() == 2 && !completedPaths.count(components[0])) {
        std::filesystem::remove(entry.path());
        std::cout << components[0] << " - Deleted 💀" << std::endl;
      }
    }
  } catch (const std::exception &e) {
    std::cerr << "Error deleting files: " << e.what() << std::endl;
  }
  std::cout << "All trajectories up-to-date!" << std::endl;
  return 0;
}

File: src/main/java/org/littletonrobotics/frc2025/subsystems/drive/trajectory/GenerateTrajectories.java
// Copyright (c) 2025 FRC 6328
// http://github.com/Mechanical-Advantage
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file at
// the root directory of this project.

package org.littletonrobotics.frc2025.subsystems.drive.trajectory;

import static org.littletonrobotics.vehicletrajectoryservice.VehicleTrajectoryServiceOuterClass.*;

import com.google.common.hash.Hashing;
import edu.wpi.first.math.util.Units;
import io.grpc.Grpc;
import io.grpc.InsecureChannelCredentials;
import java.io.*;
import java.math.RoundingMode;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.text.DecimalFormat;
import java.util.*;
import org.littletonrobotics.frc2025.Constants;
import org.littletonrobotics.frc2025.subsystems.drive.DriveConstants;
import org.littletonrobotics.vehicletrajectoryservice.VehicleTrajectoryServiceGrpc;

public class GenerateTrajectories {
  public static void main(String[] args) {
    Constants.disableHAL();

    // Create vehicle model
    VehicleModel model =
        VehicleModel.newBuilder()
            .setMass(67)
            .setMoi(5.8)
            .setVehicleLength(DriveConstants.trackWidthX)
            .setVehicleWidth(DriveConstants.trackWidthY)
            .setWheelRadius(Units.inchesToMeters(2.0))
            .setMaxWheelTorque(3.0)
            .setMaxWheelOmega(DriveConstants.maxLinearSpeed / Units.inchesToMeters(2.0) * 0.75)
            .build();

    // Connect to service
    var channel =
        Grpc.newChannelBuilder("127.0.0.1:56328", InsecureChannelCredentials.create()).build();
    var service = VehicleTrajectoryServiceGrpc.newBlockingStub(channel);

    Set<String> completedPaths = new HashSet<>();
    while (true) {
      // Get all paths
      var allPaths = DriveTrajectories.paths;
      boolean hasAllPaths = true;
      for (var supplier : DriveTrajectories.suppliedPaths) {
        var suppliedPaths = supplier.apply(completedPaths);
        if (suppliedPaths == null) {
          hasAllPaths = false;
        } else {
          allPaths.putAll(suppliedPaths);
        }
      }
      Set<String> originalKeys = new HashSet<>();
      originalKeys.addAll(allPaths.keySet());
      for (String name : originalKeys) {
        if (completedPaths.contains(name)) {
          allPaths.remove(name);
        } else {
          completedPaths.add(name);
        }
      }
      if (!hasAllPaths && allPaths.size() == 0) {
        throw new RuntimeException(
            "Invalid dependency tree. Not all supplied trajectories are available, but there are no trajectories to generate.");
      }

      // Check hashcodes
      Map<String, List<PathSegment>> pathQueue = new HashMap<>();
      for (Map.Entry<String, List<PathSegment>> entry : allPaths.entrySet()) {
        String hashCode = getHashCode(model, entry.getValue());
        File pathFile =
            Path.of("src", "main", "deploy", "trajectories", entry.getKey() + ".pathblob").toFile();
        if (pathFile.exists()) {
          try {
            InputStream fileStream = new FileInputStream(pathFile);
            Trajectory trajectory = Trajectory.parseFrom(fileStream);
            if (!trajectory.getHashCode().equals(hashCode)) {
              pathQueue.put(entry.getKey(), entry.getValue());
            }
          } catch (IOException e) {
            e.printStackTrace();
          }
        } else {
          pathQueue.put(entry.getKey(), entry.getValue());
        }
      }

      // Generate trajectories
      String generateEmptyFlag = System.getenv("GENERATE_EMPTY_DRIVE_TRAJECTORIES");
      boolean generateEmpty = generateEmptyFlag != null && generateEmptyFlag.length() > 0;
      for (Map.Entry<String, List<PathSegment>> entry : pathQueue.entrySet()) {
        Trajectory trajectory;
        System.out.print(entry.getKey() + " - Generating 💭");
        double startTime = System.currentTimeMillis();
        if (generateEmpty) {
          trajectory =
              Trajectory.newBuilder()
                  .addStates(TimestampedVehicleState.newBuilder().build())
                  .build();
        } else {
          // Use service for generation
          PathRequest request =
              PathRequest.newBuilder().setModel(model).addAllSegments(entry.getValue()).build();
          TrajectoryResponse response = service.generateTrajectory(request);
          String error = response.getError().getReason();
          if (error.length() > 0) {
            System.err.println(
                "Got error response for trajectory \"" + entry.getKey() + "\": " + error);
            System.exit(1);
          }
          trajectory =
              response.getTrajectory().toBuilder()
                  .setHashCode(getHashCode(model, entry.getValue()))
                  .build();
        }
        File pathFile =
            Path.of("src", "main", "deploy", "trajectories", entry.getKey() + ".pathblob").toFile();
        try {
          OutputStream fileStream = new FileOutputStream(pathFile);
          trajectory.writeTo(fileStream);
          double endTime = System.currentTimeMillis();
          System.out.println(
              "\r"
                  + entry.getKey()
                  + " - Finished in "
                  + Math.round((endTime - startTime) / 100.0) / 10.0
                  + " secs ✅");
        } catch (IOException e) {
          System.out.println("\r" + entry.getKey() + " - FAILED ⛔");
          e.printStackTrace();
        }
      }

      // Exit if all trajectories ready
      if (hasAllPaths) {
        break;
      }
    }

    // Delete old trajectories
    try {
      Files.list(Path.of("src", "main", "deploy", "trajectories"))
          .forEach(
              (path) -> {
                String filename = path.getFileName().toString();
                if (!filename.endsWith(".pathblob")) return;
                String[] components = filename.split("\\.");
                if (components.length == 2 && !completedPaths.contains(components[0])) {
                  path.toFile().delete();
                  System.out.println(components[0] + " - Deleted 💀");
                }
              });
    } catch (IOException e) {
      e.printStackTrace();
    }
    System.out.println("All trajectories up-to-date!");
  }

  // create a hashcode for the vehicle model and path segments

  private static String getHashCode(VehicleModel model, List<PathSegment> segements) {
    StringBuilder hashString = new StringBuilder();

    DecimalFormat format = new DecimalFormat("#.000000");
    format.setRoundingMode(RoundingMode.HALF_DOWN);

    hashString.append(format.format(model.getMass()));
    hashString.append(format.format(model.getMoi()));
    hashString.append(format.format(model.getVehicleLength()));
    hashString.append(format.format(model.getVehicleWidth()));
    hashString.append(format.format(model.getWheelRadius()));
    hashString.append(format.format(model.getMaxWheelTorque()));
    hashString.append(format.format(model.getMaxWheelOmega()));

    for (PathSegment segment : segements) {
      for (Waypoint waypoint : segment.getWaypointsList()) {
        hashString.append(format.format(waypoint.getX()));
        hashString.append(format.format(waypoint.getY()));
        if (waypoint.hasHeadingConstraint()) {
          hashString.append(format.format(waypoint.getHeadingConstraint()));
        }

        if (waypoint.hasSamples()) {
          hashString.append(waypoint.getSamples());
        }

        switch (waypoint.getVelocityConstraintCase()) {
          case ZERO_VELOCITY -> {
            hashString.append(format.format(0));
          }
          case VEHICLE_VELOCITY -> {
            hashString.append(format.format(waypoint.getVehicleVelocity().getVx()));
            hashString.append(format.format(waypoint.getVehicleVelocity().getVy()));
            hashString.append(format.format(waypoint.getVehicleVelocity().getOmega()));
          }
          case VELOCITYCONSTRAINT_NOT_SET -> {}
        }
      }

      if (segment.hasMaxVelocity()) {
        hashString.append(format.format(segment.getMaxVelocity()));
      }
      if (segment.hasMaxOmega()) {
        hashString.append(format.format(segment.getMaxOmega()));
      }

      hashString.append(segment.getStraightLine());
    }

    return Hashing.sha256().hashString(hashString, StandardCharsets.UTF_8).toString();
  }
}
